cloud_functions/gcs_event_based_ingest/README.md

```diff
@@ -0,0 +1,248 @@
+a special `_config/` prefix at the root of the bucket and/or under the dataset
+and/or table and/or under the partition prefixes.
+
+For example if you have the following files:
+```text
+gs://${INGESTION_BUCKET}/_config/load.json`
+gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`
+gs://${INGESTION_BUCKET}/${BQ_DATASET}/${BQ_TABLE_NAME}/_config/load.json
+```
+The json objects will be merged where key conflicts are resolved by config in
+the closest directory to the data.
+If the files contents were like this:
+`gs://${INGESTION_BUCKET}/_config/load.json`:
+```json
+{
+    "sourceFormat": "CSV",
+    "writeDisposition": "WRITE_APPEND",
+    "schemaUpdateOptions": ["ALLOW_FILED_RELAXATION"]
+}
+```
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`:
+```json
+{
+    "sourceFormat": "AVRO"
+}
+```
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`:
+```json
+{
+    "writeDisposition": "WRTITE_TRUNCATE"
+}
+```
+
+The result of merging these would be:
+```json
+{
+    "sourceFormat": "AVRO",
+    "writeDisposition": "WRITE_TRUNCATE",
+    "schemaUpdateOptions": ["ALLOW_FILED_RELAXATION"]
+}
+```
+This configuration system gives us the ability to DRY up common defaults but
+override them at whatever level is appropriate as new cases come up.
+
+#### Transformation SQL
+In some cases we may need to perform transformations on the files in GCS
+before they can be loaded to BigQuery. This is handled by query on an
+temporary external table over the GCS objects as a proxy for load job.
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/${BQ_TABLE_NAME}/_config/bq_transform.sql`
+
+Note, external queries will consume query slots from this project's reservation
+or count towards your on-demand billing. They will _not_ use free tie load slots.
+
+Note, that the query should select from a `temp_ext` which will be a temporary
+external table configured on the fly by the Cloud Function.
+The query must handle the logic for inserting into the destination table.
+This means it should use BigQuery DML to either `INSERT` or `MERGE` into the
+destination table.
+For example:
+```sql
+INSERT {dest_dataset}.{dest_table}
+SELECT * FROM temp_ext
+```
+Note that `{dest_dataset}` and `{dest_table}` can be used to inject the dataset
+and table inferred from the GCS path.
+
+
+The query will be run with the appropriate [external table definitions](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#ExternalDataConfiguration)
+defined in:
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/${BQ_TABLE_NAME}/_config/external.json`
+If this file is missing the external table will be assumed to be `PARQUET` format.
+
+### Partitions
+
+#### Partition Table Decorators
+Note that if the directory immediately before the triggering successfile starts with
+a `$` it will be treated as a BigQuery Partition decorator for the destination table.
+
+This means for:
+```text
+gs://${BUCKET}/foo/bar/$20201026/_SUCCESS
```

Comment:
Why not just default this to the hive partitioning layout?

Reply:
My thought is this default supports date, timestamp and integer range partition decorators and is explicit / matches BQ target syntax.
The hive partitioning is useful only to those following a supported folder structure.
In a future PR I can add more detailed docs and an integration test for how we'd suggest using this function with native BQ support for hive partitioning by specifying it in load.json at the table level.

The way the docs for BQ hive partitioning reads is that you can provide `dt=/val=` however most often in the wild (e.g. my current customer) I see data lakes of partitioned data in `table/yyyy/MM/dd/hh/` which is [unsupported by bq hive partitioning](https://cloud.google.com/bigquery/docs/hive-partitioned-loads-gcs#unsupported_data_layouts) because partition key name is not encoded in the path. Even for `yr=yyyy/m=MM/d=dd/hr=hh` (ala my first ad tech customer) this is "supported" but all of a sudden you can only partition on fields `yr`, `m`, `d`, or  `h` rather than the whole date time (which defeats the purpose). Often times there are other ETL process / consumers who need the data in this exact layout so the customer is not that flexible to change.

For these customers I'd recommend using external query loads and define a small udf to parse your org's partitioning / naming convention and constructing the timestamp / date time field you want to partition on. Alternatively they can fork this function for those use cases and add some extra parsing step on the partition extracted by regex to produce the partition decorator they desire for their destination table ref.